gitpod-io/gitpod
# [local-app][code] api to observe tunnels

#### What it does

This PR implements the tunnel provider in VS Code Web. The Local App exposes the api to stream known tunnels via grpcweb which VS Code web is listening to update internal state: https://github.com/gitpod-io/vscode/commit/c4646a76fb5abdd6e0b5a96efb1f74f5b3ed209d

This PR does not implement the tunnel provider VS Code Desktop, but the approach would be quite similar but by using VS Code Extension APIs.

<a href="https://www.loom.com/share/18a6360d06f04239a92274a2a6bb4231"><img style="max-width:300px;" src="https://cdn.loom.com/sessions/thumbnails/18a6360d06f04239a92274a2a6bb4231-with-play.gif"> </a>

#### How to test

- Start the workspace and check that ports are working as usual: https://ak-vscode-tunnel-provider.staging.gitpod-dev.com#https://github.com/akosyakov/parcel-demo
- Start the app
```bash
# mac
curl -OL https://ak-vscode-tunnel-provider.staging.gitpod-dev.com/static/bin/gitpod-local-companion-darwin
# linux
# curl -OL https://ak-vscode-tunnel-provider.staging.gitpod-dev.com/static/bin/gitpod-local-companion-linux
# windows
# curl -OL https://ak-vscode-tunnel-provider.staging.gitpod-dev.com/static/bin/gitpod-local-companion-windows

chmod +x ./gitpod-local-companion-darwin
GITPOD_HOST=https://ak-vscode-tunnel-provider.staging.gitpod-dev.com ./gitpod-local-companion-darwin --mock-keyring
```
- Reload the page and check that:
  - now localhost links are resolved to localhost, try to follow a link from the terminal
  - ports view display there port is opened on localhost or all interfaces
  - ports view allow to switch tunnel visibility
  - running open preview or open browser actually open on localhost
- Stop the app and check again that ports are working as before (not aware of tunnels).

Bonus:
- Use `Developer: Enable Tunneled Ports View` to get a view to control tunnels.
- Start the local app api endpoint on another port with `--api-port=5000`, and use command `Developer: Connect to Local Gitpod App` to connect using such port.
